### DISCUSSION TITLE
Right way for multiple Upload inputs

### DISCUSSION THREAD

**Discussion by artemklushin** [orchidsoftware/platform#1709] [Q&A]
Hello.

I want to load a file for each app language and return its depending on request language.
I read documentation, but can't understand how to set multiple Upload inputs.
What is right way for settings three Upload inputs on EditScreen?

**Comment by artemklushin** [comment]
My work example

In EditScreen

```php
Upload::make('data.filesRu')
   ->maxFiles(1)
   ->groups('ru')
   ->title('Файл (рус)')
   ->vertical(),
Upload::make('data.filesEn')
   ->maxFiles(1)
   ->groups('en')
   ->title('Файл (англ)')
   ->vertical(),
Upload::make('data.filesZh')
   ->maxFiles(1)
   ->groups('zh')
   ->title('Файл (китай)')
   ->vertical(),
```

In Eloquent model
```php
    public function filesRu()
    {
        return $this->attachment('ru');
    }

    public function filesEn()
    {
        return $this->attachment('en');
    }

    public function filesZh()
    {
        return $this->attachment('zh');
    }
```

When I save model
```php
$model->attachment('ru')->syncWithoutDetaching($properties['filesRu']);
$model->attachment('en')->syncWithoutDetaching($properties['filesEn']);
$model->attachment('zh')->syncWithoutDetaching($properties['filesZh']);
```

Is it right way?

**Reply by tabuna** [reply]
If you do not need grouping (e.g., documents, photos, etc.), then this is not a bad way.

**Reply by wsdnathan** [reply]
@tabuna I'm still confused, for his example does this require text fields filesRu, filesEn, filesZh to additionally exist in the database table?

Even though the attachmentable records exist, I cannot seem to get query() to display anything with Upload::make unless I also set in the model:
> protected $fillable = ['filesRu', 'filesEn', 'filesZh'];
> protected $casts = ['filesRu' => 'array', 'filesEn' => 'array', 'filesZh' => 'array'];

I cannot seem to get it to work without that. Is it possible to use attachmentable for the relation without adding these extra text fields? Maybe if @artemklushin did that then it would be working for him? It is not in the documentation.

**Reply by wsdnathan** [reply]
Two more points of confusion for me:
1. If you delete a file via Upload UI and save, the attachmentable record is not removed. (edit: it will remove if you use ->sync() instead)
2. If you delete files one at a time via Upload UI it works, but if you try to delete the last file it does not work because `$request->input('data.filesEn', [])` for example is empty.

**Reply by tabuna** [reply]
Let me give you an example that will cover 90% of the requests. We have a regular relationship that we added with the trait.
This can be seen in the documentation and in a step-by-step example of working with [attachments](https://orchid.software/en/docs/quickstart-files/). We have a division into groups:

```php
Upload::make('model.attachment')
    ->title('Photos')
   ->groups('images'),

Upload::make('model.attachment')
    ->title('Documents')
   ->groups('docs'),
```

This will allow one relationship to be used, stored, and received. Example:

```php
// save 
$model->attachment()->syncWithoutDetaching(
    $request->input('model.attachment', [])
);

// get all models for relations
$model->load('attachment');


// get only group
$model->attachment->where('group', 'docs')
```

Treat it like a regular relationship (Blog post and comments)

As for saving without relation, it is also possible, but for this, you need to use a separate column and cast in the model

**Comment by wsdnathan** [comment]
I am wondering the same thing. What is the correct way to define attachment relationships? Does the "attachmentable" table need to be set and what is the purpose of this?

**Reply by wsdnathan** [reply]
Oh I see, I was looking at this documentation:
https://orchid.software/en/docs/attachments/
However the information is at the bottom of this page:
https://orchid.software/en/docs/quickstart-files/

**Comment by modmatrix** [comment]
I finally got this working. (This might not be the best way.)

First of all, make sure to use the Attachable trait in your model.

I defined my Upload field like this:

```
Upload::make('device.attachment')
    ->storage('device_files')
    ->acceptedFiles('application/pdf')
    ->groups('files'),
```

And in my save handler:

```
$attachment_ids = $request->input('device.attachment', []);
$orphaned_attachments = $device->attachment->whereIn('group', ['files'])->whereNotIn('id', $attachment_ids);

$device->attachment()->sync(
    $attachment_ids
);

foreach ($orphaned_attachments as $orphaned_attachment) {
    $orphaned_attachment->delete();
}
```
